### PR TITLE
Support extend a custom auth type

### PR DESCRIPTION
**Pull request by maobaolong**
### What changes are proposed in this pull request?

Support extend a custom auth type.

### Why are the changes needed?

With this PR, we can use custom auth type for a new auth implementation, otherwise, we have to add a new auth type to the AuthType enum and proto. 

### Does this PR introduce any user facing changes?

```
alluxio.security.authentication.custom.provider.class=
alluxio.security.authentication.custom.sasl.client.handler.class=
alluxio.security.authentication.custom.sasl.server.handler.class=
```

### CODE REVIEW

**Comment by ggezer** [comment]
This change breaks the existing custom authentication support.

Did you check below property key already and see if it serves whatever you're planning to achieve with custom sasl-handlers?

  public static final PropertyKey SECURITY_AUTHENTICATION_CUSTOM_PROVIDER_CLASS =
      new Builder(Name.SECURITY_AUTHENTICATION_CUSTOM_PROVIDER_CLASS)
          .setDescription("The class to provide customized authentication implementation, "
              + "when alluxio.security.authentication.type is set to CUSTOM. It must "
              + "implement the interface "
              + "'alluxio.security.authentication.AuthenticationProvider'.")
          .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
          .setScope(Scope.ALL)
          .build();

**Comment by maobaolong** [comment]
@ggezer Thanks for your reply. 

```
alluxio.security.authentication.custom.provider.class=
alluxio.security.authentication.custom.sasl.client.handler.class=
alluxio.security.authentication.custom.sasl.server.handler.class=
```

The three key about custom auth type here, the first one is the existing custom provider, but we have two implementations  else, one extended from `SaslServerHandler`, the other extended from `SaslClientHandlerPlain`, so that we have to create this PR to extend custom auth type to let our implementation works.

**Comment by jiacheliu3** [comment]
@yyongycy Could you take a look at this PR? Thanks!

File: core/common/src/main/java/alluxio/security/authentication/ChannelAuthenticator.java [hunk 1]
```diff
@@ -164,9 +164,21 @@ private SaslClientHandler createSaslClientHandler(GrpcServerAddress serverAddres
       ChannelAuthenticationScheme authScheme, Subject subject) throws UnauthenticatedException {
     switch (authScheme) {
       case SIMPLE:
-      case CUSTOM:
         return new alluxio.security.authentication.plain.SaslClientHandlerPlain(mParentSubject,
             mConfiguration);
+      case CUSTOM:
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply core/common/src/main/java/alluxio/security/authentication/ChannelAuthenticator.java:169]: Done.

File: core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java [hunk 2]
```diff
@@ -128,8 +128,21 @@ public SaslServerHandler createSaslHandler(ChannelAuthenticationScheme authSchem
       throws SaslException {
     switch (authScheme) {
       case SIMPLE:
-      case CUSTOM:
         return new SaslServerHandlerPlain(mHostName, mConfiguration, mImpersonationAuthenticator);
+      case CUSTOM:
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java:132]: Done by previous commit

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 3]
```diff
@@ -5419,6 +5419,17 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.ALL)
           .build();
+  public static final PropertyKey SECURITY_AUTHENTICATION_CUSTOM_SASL_CLIENT_HANDLER_CLASS =
+      new Builder(Name.SECURITY_AUTHENTICATION_CUSTOM_SASL_CLIENT_HANDLER_CLASS)
+          .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:5426]: Done

**Comment by maobaolong** [comment]
@yyongycy Thanks for your suggestion, I push a new commit, PTAL.

**Review by yyongycy — APPROVED** [review]
LGTM

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 4]
```diff
@@ -5419,6 +5419,21 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.ALL)
           .build();
+  public static final PropertyKey SECURITY_AUTHENTICATION_CUSTOM_SASL_CLIENT_HANDLER_CLASS =
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:5422]:
```suggestion
  public static final PropertyKey SECURITY_AUTHENTICATION_CUSTOM_SASL_CLIENT_CLASS =
```

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 5]
```diff
@@ -5419,6 +5419,21 @@ public String toString() {
...
+      new Builder(Name.SECURITY_AUTHENTICATION_CUSTOM_SASL_CLIENT_HANDLER_CLASS)
+          .setDescription(
+              "The class to provide a custom security authentication sasl client handler.")
+          .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
+          .setScope(Scope.CLIENT)
+          .build();
+  public static final PropertyKey SECURITY_AUTHENTICATION_CUSTOM_SASL_SERVER_HANDLER_CLASS =
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:5429]:
```suggestion
  public static final PropertyKey SECURITY_AUTHENTICATION_CUSTOM_SASL_SERVER_CLASS =
```

**Review by jiacheliu3 — APPROVED** [review]
LGTM with a minor comment on the naming. I feel the "handler" in the name is not necessary. A "client handler" is still on the client side so let's just call it "client" to avoid confusion. Same for the server side equivalent.

**Comment by maobaolong** [comment]
@jiacheliu3 Thanks for your review, accepted your suggestion, and push a new extra commit to solve the other naming issue.

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 6]
```diff
@@ -5419,6 +5419,21 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.ALL)
           .build();
+  public static final PropertyKey SECURITY_AUTHENTICATION_CUSTOM_SASL_CLIENT_CLASS =
+      new Builder(Name.SECURITY_AUTHENTICATION_CUSTOM_SASL_CLIENT_CLASS)
+          .setDescription(
+              "The class to provide a custom security authentication sasl client handler.")
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:5425]: Update the descriptions according to the new names?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Please update the relevant descriptions. Otherwise LGTM, thanks!

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 6 above)
Comment (maobaolong) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: Thanks for the reminding.

**Comment by maobaolong** [comment]
@jiacheliu3 Thanks for reminding me, I push a new commit to address your comment. PTAL

**Pull request closed without merging**